File: paper_artifacts/experiments/llr9/artifacts/sources/kimi27sglang-c-skills/wf_triangular__submitted.c
#include <stdint.h>
#include <omp.h>

void wf_triangular_fp64(double *restrict a, const int64_t LEN_2D) {
  const int64_t N = LEN_2D;

  if (N < 512) {
    for (int64_t i = 1; i < N; ++i) {
      for (int64_t j = i; j < N; ++j) {
        a[i * N + j] = a[i * N + j] + a[(i - 1) * N + j] + a[i * N + (j - 1)];
      }
    }
    return;
  }

  const int64_t B = 30;
  const int64_t NB = (N + B - 1) / B;

  #pragma omp parallel
  {
    for (int64_t db = 0; db <= 2 * NB - 2; ++db) {
      int64_t bi0 = db - (NB - 1);
      if (bi0 < 0) bi0 = 0;
      int64_t bi1 = db / 2;
      if (bi1 >= NB) bi1 = NB - 1;

      #pragma omp for schedule(static)
      for (int64_t bi = bi0; bi <= bi1; ++bi) {
        const int64_t bj = db - bi;
        const int64_t i0 = bi * B;
        const int64_t i1 = (bi + 1) * B - 1 < N ? (bi + 1) * B - 1 : N - 1;
        const int64_t j0 = bj * B;
        const int64_t j1 = (bj + 1) * B - 1 < N ? (bj + 1) * B - 1 : N - 1;
        const int64_t istart = i0 < 1 ? 1 : i0;

        if (bj == bi) {
          for (int64_t i = istart; i <= i1; ++i) {
            double *restrict row = a + i * N;
            const double *restrict nrow = a + (i - 1) * N;
            for (int64_t j = i; j <= j1; ++j) {
              row[j] = row[j] + nrow[j] + row[j - 1];
            }
          }
        } else {
          for (int64_t i = istart; i <= i1; ++i) {
            double *restrict row = a + i * N;
            const double *restrict nrow = a + (i - 1) * N;
            for (int64_t j = j0; j <= j1; ++j) {
              row[j] = row[j] + nrow[j] + row[j - 1];
            }
          }
        }
      }
    }
  }
}

File: paper_artifacts/experiments/llr9/artifacts/sources/oss120b-fortran/wf_triangular__submitted.f90
subroutine wf_triangular_fp64(a, LEN_2D, workspace, workspace_bytes) bind(C, name="wf_triangular_fp64")
    use iso_c_binding
    implicit none
    real(c_double), intent(inout) :: a(*)
    integer(c_int64_t), value :: LEN_2D
    type(c_ptr), value :: workspace
    integer(c_int64_t), value :: workspace_bytes
    integer :: i, j, k
    integer :: i_start, i_end
    integer :: idx
    integer :: len2
    ! Convert LEN_2D (C int64) to default integer for indexing.
    len2 = LEN_2D
    ! Wavefront parallelism using row-major indexing on the flattened array.
    !$omp parallel private(i, j, k, i_start, i_end, idx)
    do k = 4, 2 * len2
        i_start = max(2, k - len2)
        i_end   = min(len2, k / 2)
        !$omp do schedule(static)
        do i = i_start, i_end
            j = k - i
            idx = (i - 1) * len2 + j
            a(idx) = a(idx) + a(idx - len2) + a(idx - 1)
        end do
        !$omp end do
    end do
    !$omp end parallel
end subroutine wf_triangular_fp64
